backend/status/status_source.go

```diff
@@ -0,0 +1,213 @@
+package status
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+	"os"
+	"syscall"
+
+	grpccodes "google.golang.org/grpc/codes"
+	grpcstatus "google.golang.org/grpc/status"
+)
+
+// Source type defines the status source.
+type Source string
+
+const (
+	// SourcePlugin status originates from plugin.
+	SourcePlugin Source = "plugin"
+
+	// SourceDownstream status originates from downstream service.
+	SourceDownstream Source = "downstream"
+
+	// DefaultSource is the default [Source] that should be used when it is not explicitly set.
+	DefaultSource Source = SourcePlugin
+)
+
+// IsValid return true if es is [SourceDownstream] or [SourcePlugin].
+func (s Source) IsValid() bool {
+	return s == SourceDownstream || s == SourcePlugin
+}
+
+// String returns the string representation of s. If s is not valid, [DefaultSource] is returned.
+func (s Source) String() string {
+	if !s.IsValid() {
+		return string(DefaultSource)
+	}
+
+	return string(s)
+}
+
+// SourceFromHTTPError returns a [Source] based on provided error.
+func SourceFromHTTPError(err error) Source {
+	if IsDownstreamHTTPError(err) {
+		return SourceDownstream
+	}
+	return SourcePlugin
+}
+
+// ErrorSourceFromStatus returns a [Source] based on provided HTTP status code.
+func SourceFromHTTPStatus(statusCode int) Source {
+	switch statusCode {
+	case http.StatusMethodNotAllowed,
+		http.StatusNotAcceptable,
+		http.StatusPreconditionFailed,
+		http.StatusRequestEntityTooLarge,
+		http.StatusRequestHeaderFieldsTooLarge,
+		http.StatusRequestURITooLong,
+		http.StatusExpectationFailed,
+		http.StatusUpgradeRequired,
+		http.StatusRequestedRangeNotSatisfiable,
+		http.StatusNotImplemented:
+		return SourcePlugin
+	}
+
+	return SourceDownstream
+}
+
+type errorWithSourceImpl struct {
```

Comment:
Could we export  `ErrorWithSourceImpl` and then use it everywhere in backend/experimental/errorsource instead of 
```
type Error struct {
	source backend.ErrorSource
	status backend.Status

	err error
}
```

The issue is that we have couple of methods in `experimental/errorsource` package, such [Response](https://github.com/grafana/grafana-plugin-sdk-go/blob/main/experimental/errorsource/error_source.go#L66) that check the error source and create response based on source. But this is going to fail with `errorWithSourceImpl`   

This issue was found by @njvrzm has found when testing my PR with elasticsearch.

Reply:
Or deprecate those experimental methods and create them in here instead.

Reply:
Given https://github.com/grafana/grafana-plugin-sdk-go/issues/1105 let's handle this in a follow up.

Reply:
👍